File: Algorithms/Sorting/shell_sort.py
#!/usr/bin/python3

def shell_sort(seq):
    n = len(seq)
    gap = n // 2
    while gap > 0:
        i = gap
        while i < n:
            cur = seq[i]
            j = i
            while j >= gap and seq[j - gap] > cur:
                seq[j] = seq[j - gap]
                j -= gap
                
            seq[j] = cur
            i += 1

        gap //= 2

    return seq

seq = [3, 5, 2, 4, 1]
print(shell_sort(seq))

File: Algorithms/Sorting/shell_sort.rs
fn shell_sort(seq: &mut [u8; 5]) {
    let mut gap: usize = seq.len() / 2;
    while gap > 0 {
        for i in gap..seq.len() {
            let cur: u8 = seq[i];
            let mut j: usize = i;
            while j >= gap && seq[j - gap] > cur {
                seq[j] = seq[j - gap];
                j -= gap;
            }
            seq[j] = cur;
        }

        gap /= 2;
    }
}

fn main() {
    let mut seq: [u8; 5] = [3, 5, 2, 4, 1];

    shell_sort(&mut seq);

    for i in &seq {
        print!("{} ", i);
    }
    print!("\n");
}
